post-processing/plot_handler.py

```diff
@@ -0,0 +1,178 @@
+                                (y_label, "@{0}_mean".format(y_column)
+                                    + ("{%0.2f}" if pd.api.types.is_float_dtype(df[y_column].dtype)
+                                       else ""))],
+                             formatters={"@{0}_mean".format(y_column): "printf"}))
+
+    # sort x-axis values in descending order (otherwise default sort is ascending)
+    reverse = False
+    if x_axis.get("sort"):
+        if x_axis["sort"] == "descending":
+            reverse = True
+
+    # sort x-axis groups by series first
+    if len(groups) > 1:
+        # get series values with their original dtype
+        # NOTE: currently not accounting for more than one series column
+        series_values = [pd.Series(x[-1], dtype=last_group_dtype).iloc[0]
+                         for x in plot.x_range.factors]
+        sorted_x_items = [x[1] for x in sorted(zip(series_values, plot.x_range.factors),
+                                               reverse=reverse, key=lambda x: x[0])]
+        plot.x_range.factors = sorted_x_items
+
+    # sort by x-axis values
+    plot.x_range.factors = sorted(plot.x_range.factors, reverse=reverse,
+                                  key=lambda x: pd.Series(x[0] if len(groups) > 1 else x,
+                                                          dtype=x_col_dtype).iloc[0])
+
+    # automatically base bar colouring on last group column
+    colour_factors = [str(x) for x in sorted(pd.Series(df[groups[-1]].unique(),
+                                                       dtype=last_group_dtype))]
+    # divide and assign colours
+    index_cmap = factor_cmap(index_group_col, palette=viridis(len(colour_factors)),
+                             factors=colour_factors, start=len(groups)-1, end=len(groups))
+    # add legend labels to data source
+    data_source = ColumnDataSource(grouped_df).data
+    legend_labels = ["{0} = {1}".format(groups[-1].replace("_", " "),
+                                        group[-1] if len(groups) > 1 else group)
+                     for group in data_source[index_group_col]]
+    data_source["legend_labels"] = legend_labels
+
+    # create legend outside plot
+    plot.add_layout(Legend(), "right")
+    # add bars
+    plot.vbar(x=index_group_col, top="{0}_mean".format(y_column), width=0.9, source=data_source,
+              line_color="white", fill_color=index_cmap, legend_group="legend_labels", hover_alpha=0.9)
+    # add labels
+    plot.xaxis.axis_label = x_label
+    plot.yaxis.axis_label = y_label
+    # remove x-axis group ticks
+    plot.xaxis.major_tick_line_color = None
+    plot.xaxis.major_label_text_font_size = "0pt"
+    # FIXME: add this as a config option at some point
+    # plot.xaxis.group_label_orientation = "vertical"
+    # adjust font size
+    plot.title.text_font_size = "15pt"
+
+    # get label values with their original dtype
+    label_values = [pd.Series(x.label["value"].split("=")[1].strip(), dtype=last_group_dtype).iloc[0]
```

Comment:
Specifically, I think changing `x.label.value` to `x.label["value"]` is needed to work with Bokeh v2.4.3, which is apparently the only Streamlit-compatible version. I haven't had the chance to test that this doesn't break anything with the current Bokeh v3.1.1, which is the version I was using until I put together the demo for #259.

Reply:
Good to keep a note of this! But don't change it here, do it in the PR that will introduce the need for it, which will be the one for streamlit.